Split remote cluster watching from reversetunnel.AgentPool

awly opened the pull request:
Separating the responsibilities:
- AgentPool takes a proxy (or LB) endpoint and manages a pool of agents
  for it (each agent is a tunnel to a unique proxy process behind the
  endpoint)
- RemoteClusterWatcher polls the auth server for a list of trusted
  clusters and manages a set of AgentPools, one for each trusted cluster

Previously, AgentPool did both of the above.

Also, bundling some cleanup in the area:
- better error when dialing through tunnel and directly both fail
- rename RemoteKubeProxy to LocalKubernetes to better reflect the
  meaning
- remove some dead code and simplify config structs

Commit:
Split remote cluster watching from reversetunnel.AgentPool

Separating the responsibilities:
- AgentPool takes a proxy (or LB) endpoint and manages a pool of agents
  for it (each agent is a tunnel to a unique proxy process behind the
  endpoint)
- RemoteClusterTunnelManager polls the auth server for a list of trusted
  clusters and manages a set of AgentPools, one for each trusted cluster

Previously, AgentPool did both of the above.

Also, bundling some cleanup in the area:
- better error when dialing through tunnel and directly both fail
- rename RemoteKubeProxy to LocalKubernetes to better reflect the
  meaning
- remove some dead code and simplify config structs

fspmarshall commented on lib/reversetunnel/watcher.go:
```diff
@@ -0,0 +1,241 @@
+//
+// Note: ReverseTunnel resources on the auth server represent the desired
+// tunnels, not actually active ones.
+type RemoteClusterWatcher struct {
```
Nit: I'm a bit hesitant about this name, just because this type is doing a lot more than what a typical `*Watcher` type does in teleport.  Would it be possible to use something a bit more descriptive (e.g. `RemoteClusterTunneler`)?

awly replied:
Renamed to `RemoteClusterTunnelManager` :)

fspmarshall reviewed (approved):
Since `AgentPool` no longer handles multiple separate endpoints, a lot of the logic for `track.Tracker` and `workpool.Pool` could be similarly simplified.  Both of those types relied on internal mappings to maintain per-endpoint state, which isn't a thing anymore (the `track.Key` type can be factored out completely).

Of course, refactoring both those types and their tests is a lot of work.  Just adding some `TODO`s indicating that that there is room for future simplification seems fine. Simplify in stages.

Also, nice work!  I'm excited about the prospect of a more understandable/maintainable `reversetunnel` package.

Commit:
reversetunnel: factor out track.Key

ClusterName is the same for all Agents in an AgentPool. track.Tracker
needs to only track proxy addresses.

awly commented:
@fspmarshall good point, I was able to factor out `track.Key` and replace it with `utils.NetAddr`.
`AgentPool` still tracks multiple endpoints (multiple proxies behind an LB) but at least they all belong to the same cluster.

fspmarshall commented:
> `AgentPool` still tracks multiple endpoints (multiple proxies behind an LB) but at least they all belong to the same cluster.

IIUC, all proxies will have the same address since they are all behind the same LB, so even the address key isn't necessary (i.e. `AgentPool.agents` can be changed from `map[utils.NetAddr][]*Agent` to `[]*Agent`,  `Tracker.sets` can be changed from `map[utils.NetAddr]*proxySet` to `*proxySet`, etc...).

That being said... maybe its worth preserving this logic as-is because we have been talking about using "sharded" proxies :thinking:
